### PR TITLE
internal/ethapi: add more testcase for GetTransactionReceipt

### PR DESCRIPTION
**Pull request by jsvisa**

### CODE REVIEW

**Review by lightclient — COMMENTED** [review]
This test seems nice, but I feel like we should be putting our energy into adding rpc tests to https://github.com/ethereum/execution-apis

**Comment by jsvisa** [comment]
I'm going to refact the code in [internal/ethapi](internal/ethapi), so test code is first introduced.

> This test seems nice, but I feel like we should be putting our energy into adding rpc tests to https://github.com/ethereum/execution-apis

@lightclient Do you mean we should also add test cases inside https://github.com/ethereum/execution-apis/tree/main/tests/eth_getTransactionReceipt, if so, I'll file an PR in https://github.com/ethereum/execution-apis

**Comment by lightclient** [comment]
Yes exactly. You can find out more info on how to generate those tests here: https://github.com/ethereum/execution-apis/blob/main/tests/README.md

**Comment by jsvisa** [comment]
@lightclient please take another look. I'll file another two PRs in [rpctestgen](https://github.com/lightclient/rpctestgen) and [execution-api](https://github.com/ethereum/execution-apis)

File: internal/ethapi/api_test.go [hunk 1]
```diff
@@ -222,15 +220,20 @@ func newTestBackend(t *testing.T, n int, gspec *core.Genesis, generator func(i i
...
+	txlookupLimit := uint64(0)
+	chain, err := core.NewBlockChain(db, cacheConfig, gspec, nil, engine, vm.Config{}, nil, &txlookupLimit)
 	if err != nil {
 		t.Fatalf("failed to create tester chain: %v", err)
 	}
 	if n, err := chain.InsertChain(blocks); err != nil {
 		t.Fatalf("block %d: failed to insert into chain: %v", n, err)
 	}
-	backend.chain = chain
+	if n, err := chain.InsertReceiptChain(blocks, receipts, 0); err != nil {
```
Comment (lightclient) [inline internal/ethapi/api_test.go:232]: I believe `chain.InsertChain` already does this.

File: internal/ethapi/api_test.go [hunk 2]
```diff
@@ -323,7 +326,16 @@ func (b testBackend) StateAndHeaderByNumberOrHash(ctx context.Context, blockNrOr
 }
 func (b testBackend) PendingBlockAndReceipts() (*types.Block, types.Receipts) { panic("implement me") }
 func (b testBackend) GetReceipts(ctx context.Context, hash common.Hash) (types.Receipts, error) {
-	panic("implement me")
+	number := rawdb.ReadHeaderNumber(b.db, hash)
+	if number == nil {
+		return nil, nil
+	}
+	header, err := b.HeaderByNumber(ctx, rpc.BlockNumber(*number))
```
Comment (lightclient) [inline internal/ethapi/api_test.go:333]: What is the need for this check?

File: internal/ethapi/api_test.go [hunk 3]
```diff
@@ -990,3 +1003,153 @@ func TestRPCGetBlockOrHeader(t *testing.T) {
...
+		if err != nil {
+			t.Errorf("failed to get block: %d err: %v", i, err)
+		}
+		blockHashes[i] = header.Hash()
+	}
+
+	var testSuite = []struct {
+		txHash    common.Hash
+		want      string
+		expectErr error
```
Comment (lightclient) [inline internal/ethapi/api_test.go:1093]: If this isn't used, I wouldn't include it.

File: internal/ethapi/api_test.go [hunk 4]
```diff
@@ -990,3 +1003,153 @@ func TestRPCGetBlockOrHeader(t *testing.T) {
...
+		if err != nil {
+			t.Errorf("test %d: want no error, have %v", i, err)
+			continue
+		}
+		data, err := json.Marshal(result)
+		if err != nil {
+			t.Errorf("test %d: json marshal error", i)
+			continue
+		}
+		want, have := tt.want, string(data)
```
Comment (lightclient) [inline internal/ethapi/api_test.go:1152]: Seems like you could just rename `data` to `have` and get rid of this line.

File: internal/ethapi/api_test.go (hunk 1 above)
Comment (jsvisa) [reply internal/ethapi/api_test.go:232]: sounds good.

File: internal/ethapi/api_test.go (hunk 4 above)
Comment (jsvisa) [reply internal/ethapi/api_test.go:1152]: Ahha, then I need to type two times for `tt.want` and `string(have)` in 
 this line https://github.com/ethereum/go-ethereum/blob/389161cea069ce83c6aed6243e09feb9af6817b8/internal/ethapi/api_test.go#L1149

File: internal/ethapi/api_test.go (hunk 2 above)
Comment (jsvisa) [reply internal/ethapi/api_test.go]: Seems this check is redundant.

btw, I've rewritten those lines into `HeaderByHash`(-4 cloc), and in that we need to check the header nullable

**Review by holiman — APPROVED** [review]

**Pull request merged**